rocks-lib/src/progress.rs

```diff
@@ -1,5 +1,39 @@
 use std::{borrow::Cow, time::Duration};
 
+mod private {
+    pub trait HasProgress {}
+}
+
+impl private::HasProgress for () {}
+impl private::HasProgress for MultiProgress {}
+impl private::HasProgress for ProgressBar {}
+
+pub trait HasProgress: private::HasProgress {}
+impl<P: private::HasProgress> HasProgress for P {}
+
+#[derive(Clone)]
+pub enum Progress<T: HasProgress> {
+    NoProgress,
+    Progress(T),
+}
```

Comment:
question: Can this be a newtype wrapper around an `Option<T>`?

Reply:
I've been thinking of implementing `Deref` to an Option<T>, but I'm not sure we can uphold proper type constraint rules that way. I prefer having this as an enum because we can simply supply a `NoProgress` and have it be much more semantic.